File: ConvertSortedListToBinarySearchTree.py
"""
109. Convert Sorted List to Binary Search Tree
Given a singly linked list where elements are sorted in ascending order, convert it to a height balanced BST.
"""

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# O(n)
# O(n)
class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        if not head:
            return None
        dummy = ListNode(-1)
        dummy.next = head
        prev, slow, fast = dummy, head, head
        while fast and fast.next:
            prev, slow, fast = prev.next, slow.next, fast.next.next
        prev.next = None
        node = TreeNode(slow.val)
        node.left = self.sortedListToBST(dummy.next)
        node.right = self.sortedListToBST(slow.next)
        return node

if __name__ == "__main__":
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    root = Solution().sortedListToBST(head)
    print(root.left.val, root.val, root.right.val)

File: Java/ConvertSortedListToBinarySearchTree.java

public class ConvertSortedListToBinarySearchTree {
    public TreeNode sortedListToBST(ListNode head) {
        if(head == null) {return null;}
        return buildTree(head);
    }
    
    private TreeNode buildTree(ListNode node) {
        ListNode dummy = new ListNode(-1);
        dummy.next = node;
        ListNode fast = node;
        ListNode slow = node;
        ListNode prev = dummy;
        while(fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            prev = prev.next;
        }
        TreeNode root = new TreeNode(slow.val);
        prev.next = null;
        if(dummy.next != null) {
            root.left = buildTree(dummy.next);
        }
        if(slow.next != null) {
            root.right = buildTree(slow.next);
        }
        return root;
    }

}
